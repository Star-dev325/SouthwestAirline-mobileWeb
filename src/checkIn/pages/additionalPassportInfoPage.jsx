// @flow
import i18n from '@swa-ui/locale';
import _ from 'lodash';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import * as CheckInActions from 'src/checkIn/actions/checkInActions';
import AdditionalPassportInfoForm from 'src/checkIn/components/additionalPassportInfoForm';
import * as AdditionalPassportInfoPageSelector from 'src/checkIn/selectors/checkInAdditionalPassportInfoPageSelector';
import { getNextPageOptions, getPassengerName, getRequestData } from 'src/checkIn/selectors/checkInFlowDataSelector';
import { cleanUpEndOfSession } from 'src/shared/actions/accountActions';
import * as DialogActions from 'src/shared/actions/dialogActions';
import { raiseSatelliteEvent } from 'src/shared/analytics/helpers/analyticsEventHelper';
import PageHeaderWithButtons from 'src/shared/components/pageHeaderWithButtons';
import CheckInErrorCode from 'src/checkIn/constants/checkInErrorCode';
import { CHECK_IN_ADDITIONAL_PASSPORT_INFO_FORM } from 'src/shared/constants/formIds';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import BrowserObject from 'src/shared/helpers/browserObject';
import { containsApiErrorCodes } from 'src/shared/helpers/errorCodesHelper';
import { hasSessionExpired } from 'src/shared/helpers/loginSessionHelper';
import { buildPathWithParamAndQuery } from 'src/shared/helpers/pathUtils';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';

import type { NextPageOptionType } from 'src/checkIn/flow-typed/checkIn.types';
import type { FormData } from 'src/shared/form/flow-typed/form.types';
import type { Push } from 'src/shared/flow-typed/shared.types';

const { location } = BrowserObject;

type Props = {
  isLoggedIn: boolean,
  passengerName: string,
  recordLocator: string,
  shouldShowSkipButton: boolean,
  nextPageOptions: NextPageOptionType,
  isLastPAX: boolean,
  requestData: Link,
  checkInSessionToken: string,
  initialFormData: ?FormData,
  documentTitles?: { [string]: string },
  showDialogFn: (*) => Promise<*>,
  hideDialogFn: () => Promise<*>,
  addAdditionalPassportInfoDocumentsFn: (Link, FormData, string) => Promise<*>,
  cleanUpEndOfSessionFn: () => void,
  params: { paxNumber: number },
  push: Push,
  goBack: (*) => void
};

export class AdditionalPassportInfoPage extends Component<Props> {
  static defaultProps = {
    recordLocator: '',
    passengerName: '',
    requestData: {},
    isLastPAX: false
  };

  componentDidMount() {
    raiseSatelliteEvent('TOOL:CHCK:additional traveler details');
  }

  _onSubmit = (formData: FormData) => {
    const { addAdditionalPassportInfoDocumentsFn, checkInSessionToken, requestData } = this.props;

    addAdditionalPassportInfoDocumentsFn(requestData, formData, checkInSessionToken)
      .then(() => {
        this._transitionToNextPax();
      })
      .catch((error) => {
        if (containsApiErrorCodes(error, CheckInErrorCode.CHECKIN_SESSION_TOKEN_EXPIRED)) {
          this._showSessionExpiredPopup();
        }
      });
  };

  _onAdditionalNavItemClick = (key: string) => {
    const {
      params: { paxNumber },
      push
    } = this.props;

    const routePathMap = {
      'destination': getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformationDestination' }),
      'permanentResidentCard': getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformationGreenCard' }),
      'visa': getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformationVisa' })
    };

    push(buildPathWithParamAndQuery(routePathMap[key], { paxNumber }));
  };

  _showSessionExpiredPopup = () => {
    const { showDialogFn } = this.props;

    showDialogFn({
      name: 'check-in-travel-documents-session-token-expired',
      title: i18n('CHECK_IN__ERRORS__SESSION_TOKEN_EXPIRED'),
      buttons: [
        {
          label: i18n('SHARED__BUTTON_TEXT__OK'),
          onClick: this._reloadPage
        }
      ]
    });
  };

  _reloadPage = () => {
    const { isLoggedIn, hideDialogFn, cleanUpEndOfSessionFn } = this.props;

    if (isLoggedIn && hasSessionExpired()) {
      cleanUpEndOfSessionFn();
    }

    hideDialogFn().then(() => location.reload());
  };

  _onCancelButtonClick = () => {
    const { showDialogFn, hideDialogFn, goBack } = this.props;

    showDialogFn({
      name: 'addition-quit-apis-flow',
      title: i18n('SHARED__PASSPORT_PAGE__CANCEL_DIALOG_TITLE'),
      message: i18n('SHARED__PASSPORT_PAGE__CANCEL_DIALOG_MESSAGE_FOR_CHECK_IN'),
      buttons: [
        {
          label: i18n('SHARED__BUTTON_TEXT__NO'),
          onClick: hideDialogFn
        },
        {
          label: i18n('SHARED__BUTTON_TEXT__YES'),
          onClick: () => {
            hideDialogFn().then(goBack);
          }
        }
      ]
    });
  };

  _onSkipButtonClick = () => {
    this._transitionToNextPax();
  };

  _transitionToNextPax = () => {
    const {
      push,
      nextPageOptions: { nextPagePath, nextPaxNumber }
    } = this.props;

    if (nextPagePath === getNormalizedRoute({ routeName: 'checkInConfirmation' })) {
      push(getNormalizedRoute({ routeName: 'checkInConfirmation' }));
    } else {
      push(buildPathWithParamAndQuery(nextPagePath, { paxNumber: +nextPaxNumber }));
    }
  };

  _getButtons = (shouldShowSkipButton: boolean) => {
    const leftButtons = [{ name: 'Cancel', onClick: this._onCancelButtonClick }];
    const rightButtons = [{ name: 'Skip', onClick: this._onSkipButtonClick }];

    return shouldShowSkipButton ? { leftButtons, rightButtons } : { leftButtons };
  };

  render() {
    const {
      passengerName,
      shouldShowSkipButton,
      isLastPAX,
      initialFormData,
      documentTitles = {},
      params: { paxNumber }
    } = this.props;
    const buttons = this._getButtons(shouldShowSkipButton);

    return (
      <div>
        <PageHeaderWithButtons title="Travel Document" {...buttons} />
        <AdditionalPassportInfoForm
          formId={`${CHECK_IN_ADDITIONAL_PASSPORT_INFO_FORM}_${paxNumber}`}
          initialFormData={initialFormData}
          isLastPAX={isLastPAX}
          documentTitles={documentTitles}
          passengerName={passengerName}
          onSubmit={this._onSubmit}
          onAdditionalNavItemClick={this._onAdditionalNavItemClick}
        />
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  isLoggedIn: _.get(state, 'app.account.isLoggedIn'),
  recordLocator: _.get(state, 'app.checkIn.checkInFlowData.recordLocator'),
  checkInSessionToken: _.get(state, 'app.checkIn.checkInFlowData.checkInSessionToken'),
  nextPageOptions: getNextPageOptions(state, props),
  documentTitles: AdditionalPassportInfoPageSelector.getDocumentTitles(state, props),
  initialFormData: AdditionalPassportInfoPageSelector.getFormData(state, props),
  shouldShowSkipButton: AdditionalPassportInfoPageSelector.shouldShowSkipButton(state),
  isLastPAX: AdditionalPassportInfoPageSelector.isLastPAX(state, props),
  passengerName: getPassengerName(state, props),
  requestData: getRequestData(state, props)
});

const mapDispatchToProps = {
  showDialogFn: DialogActions.showDialog,
  hideDialogFn: DialogActions.hideDialog,
  addAdditionalPassportInfoDocumentsFn: CheckInActions.addAdditionalPassportInfoDocuments,
  cleanUpEndOfSessionFn: cleanUpEndOfSession
};

const enhancers = _.flowRight(withConnectedReactRouter, connect(mapStateToProps, mapDispatchToProps));

export default enhancers(AdditionalPassportInfoPage);
