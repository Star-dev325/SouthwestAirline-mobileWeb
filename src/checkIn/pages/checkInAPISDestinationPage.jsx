// @flow

import _ from 'lodash';
import { connect } from 'react-redux';

import withCheckInAPISPage from 'src/checkIn/enhancers/withCheckInAPISPage';
import CheckInAPISDestinationForm from 'src/checkIn/components/checkInAPISDestinationForm';
import * as FormDataActions from 'src/shared/actions/formDataActions';
import { CHECK_IN_APIS_DESTINATION_FORM } from 'src/shared/constants/formIds';

export const transformAdditionalProps = (travelDocument: *) => {
  const destinationConfig = _.get(travelDocument, 'destinationConfig', {});

  return { destinationConfig };
};

const mapStateToProps = () => ({});

const mapDispatchToProps = {
  updateFormDataValueFn: FormDataActions.updateFormDataValue
};

const enhancers = _.flowRight(
  withCheckInAPISPage({
    nodeName: 'destination',
    formId: CHECK_IN_APIS_DESTINATION_FORM,
    transformAdditionalProps
  }),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(CheckInAPISDestinationForm);
