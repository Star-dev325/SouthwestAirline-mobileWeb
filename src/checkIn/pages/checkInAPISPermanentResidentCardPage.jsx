// @flow
import withCheckInAPISPage from 'src/checkIn/enhancers/withCheckInAPISPage';
import CheckInAPISPermanentResidentCardForm from 'src/checkIn/components/checkInAPISPermanentResidentCardForm';
import { CHECK_IN_APIS_PERMANENT_RESIDENT_CARD_FORM } from 'src/shared/constants/formIds';

export default withCheckInAPISPage({
  nodeName: 'permanentResidentCard',
  formId: CHECK_IN_APIS_PERMANENT_RESIDENT_CARD_FORM
})(CheckInAPISPermanentResidentCardForm);
