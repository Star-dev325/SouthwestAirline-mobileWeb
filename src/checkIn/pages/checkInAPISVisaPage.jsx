// @flow
import withCheckInAPISPage from 'src/checkIn/enhancers/withCheckInAPISPage';
import CheckInAPISVisaForm from 'src/checkIn/components/checkInAPISVisaForm';
import { CHECK_IN_APIS_VISA_FORM } from 'src/shared/constants/formIds';

export default withCheckInAPISPage({ nodeName: 'visa', formId: CHECK_IN_APIS_VISA_FORM })(CheckInAPISVisaForm);
