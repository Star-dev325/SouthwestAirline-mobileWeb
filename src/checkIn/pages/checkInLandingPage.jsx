// @flow

import i18n from '@swa-ui/locale';
import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import * as CheckInActions from 'src/checkIn/actions/checkInActions';
import CheckInEligibleTripsLink from 'src/checkIn/components/checkInEligibleTripsLink';
import { getTripsThatNeedToCheckIn } from 'src/checkIn/selectors/checkInLandingPageSelector';
import * as SharedActions from 'src/shared/actions/sharedActions';
import RecentTripSearchCardsList from 'src/shared/components/recentTripSearchCardsList';
import SubHeader from 'src/shared/components/subHeader';
import { CHECK_IN_RETRIEVE_RESERVATION_FORM } from 'src/shared/constants/formIds';
import withBodyClass from 'src/shared/enhancers/withBodyClass';
import withConnectedReactRouter from 'src/shared/enhancers/withConnectedReactRouter';
import withRecentTripSearches from 'src/shared/enhancers/withRecentTripSearches';
import ReservationRetrievalForm from 'src/shared/form/components/reservationRetrievalForm';
import { hasActiveSessionCookies } from 'src/shared/helpers/loginSessionHelper';

import type { CheckInViewResDetailRequestType } from 'src/checkIn/flow-typed/checkIn.types';
import type { PassengerNameRecord, Push, UpcomingTripPage } from 'src/shared/flow-typed/shared.types';

type Props = {
  fetchRecentTripSearchesFn: () => Promise<*>,
  fetchUpcomingTripsNonBlockingFn: () => Promise<*>,
  isLoggedIn: boolean,
  push: Push,
  recentTripSearches: Array<PassengerNameRecord>,
  retrieveCheckInReservationDetailsFn: (CheckInViewResDetailRequestType) => Promise<*>,
  saveRecentTripSearchFn: (passengerNameRecord?: PassengerNameRecord) => *,
  tripsThatNeedToCheckIn: Array<UpcomingTripPage>
};

export class CheckInLandingPage extends React.Component<Props> {
  componentDidMount() {
    const { fetchUpcomingTripsNonBlockingFn } = this.props;

    hasActiveSessionCookies() && fetchUpcomingTripsNonBlockingFn();
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    const userJustLoggedIn = !this.props.isLoggedIn && nextProps.isLoggedIn;

    if (userJustLoggedIn) {
      this.props.fetchUpcomingTripsNonBlockingFn();
    }
  }

  _onSubmit = (requestModel: CheckInViewResDetailRequestType) => {
    const { saveRecentTripSearchFn, fetchRecentTripSearchesFn } = this.props;

    saveRecentTripSearchFn(requestModel);
    fetchRecentTripSearchesFn();

    this.props.retrieveCheckInReservationDetailsFn(requestModel);
  };

  _isLoggedInAndHasCheckInEligibleTrips = (isLoggedIn: boolean, tripsThatNeedToCheckIn: Array<*>) =>
    isLoggedIn && tripsThatNeedToCheckIn.length > 0;

  _transitionToUpcomingTripsOrViewReservationDetailsPage = () => {
    const { tripsThatNeedToCheckIn, retrieveCheckInReservationDetailsFn, push } = this.props;

    if (tripsThatNeedToCheckIn.length === 1) {
      const trip = tripsThatNeedToCheckIn[0];

      const requestData = {
        recordLocator: trip.confirmationNumber,
        firstName: _.get(trip, '_links.checkInViewReservationPage.query.first-name'),
        lastName: _.get(trip, '_links.checkInViewReservationPage.query.last-name')
      };

      retrieveCheckInReservationDetailsFn(requestData);
    } else if (tripsThatNeedToCheckIn.length > 1) {
      push('/my-account/upcoming-trips');
    }
  };

  render() {
    const { recentTripSearches, tripsThatNeedToCheckIn, isLoggedIn } = this.props;

    return (
      <div>
        <SubHeader title="Check In" />

        {this._isLoggedInAndHasCheckInEligibleTrips(isLoggedIn, tripsThatNeedToCheckIn) && (
          <CheckInEligibleTripsLink
            numberOfCheckInEligibleTrips={tripsThatNeedToCheckIn.length}
            onClick={this._transitionToUpcomingTripsOrViewReservationDetailsPage}
          />
        )}

        <ReservationRetrievalForm formId={CHECK_IN_RETRIEVE_RESERVATION_FORM} onSubmit={this._onSubmit} />

        <p className="check-in--eligible-time-range-notice">{i18n('CHECK_IN__CHECK_IN_ELIGIBLE_TIME_RANGE_NOTICE')}</p>

        {!_.isEmpty(recentTripSearches) && (
          <RecentTripSearchCardsList recentTripSearches={recentTripSearches} onCardClick={this._onSubmit} />
        )}
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  isLoggedIn: _.get(state, 'app.account.isLoggedIn'),
  tripsThatNeedToCheckIn: getTripsThatNeedToCheckIn(state)
});

const mapDispatchToProps = {
  fetchUpcomingTripsNonBlockingFn: SharedActions.fetchUpcomingTripsNonBlocking,
  retrieveCheckInReservationDetailsFn: CheckInActions.getReserveCheckInReservation
};

const enhancers = _.flowRight(
  withBodyClass('check-in-landing-page'),
  withConnectedReactRouter,
  withRecentTripSearches('checkIn'),
  connect(mapStateToProps, mapDispatchToProps)
);

export default enhancers(CheckInLandingPage);
