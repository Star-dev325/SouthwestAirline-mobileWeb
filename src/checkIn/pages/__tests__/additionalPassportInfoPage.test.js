jest.mock('src/shared/helpers/browserObject', () => ({ location: { pathname: '/check-in', reload: jest.fn() } }));

import '@testing-library/jest-dom/extend-expect';
import { fireEvent, waitFor } from '@testing-library/react';
import * as CheckInActions from 'src/checkIn/actions/checkInActions';
import { AdditionalPassportInfoPage } from 'src/checkIn/pages/additionalPassportInfoPage';
import * as DialogActions from 'src/shared/actions/dialogActions';
import BrowserObject from 'src/shared/helpers/browserObject';
import * as LoginSessionHelper from 'src/shared/helpers/loginSessionHelper';
import { getNormalizedRoute } from 'src/shared/helpers/urlHelper';
import * as AppSelector from 'src/shared/selectors/appSelector';
import { integrationRender } from 'test/unit/helpers/testingLibraryUtils';

describe('AdditionalPassportInfoPage', () => {
  let addAdditionalPassportInfoDocumentsStub;
  let cleanUpEndOfSessionFnStub;
  let goBackStub;
  let hideDialogStub;
  let prefillNextPaxInfoStub;
  let prepareTravelDocumentStub;
  let pushStub;
  let saveDestinationForAllStub;
  let showDialogStub;

  beforeEach(() => {
    addAdditionalPassportInfoDocumentsStub = jest.spyOn(CheckInActions, 'addAdditionalPassportInfoDocuments')
      .mockResolvedValue('done');
    jest.spyOn(AppSelector, 'getCurrentAppFlow').mockReturnValueOnce('check-in');
    cleanUpEndOfSessionFnStub = jest.fn();
    goBackStub = jest.fn();
    hideDialogStub = jest.fn(DialogActions, 'hideDialog').mockResolvedValue('done');
    prefillNextPaxInfoStub = jest.fn();
    prepareTravelDocumentStub = jest.fn();
    pushStub = jest.fn();
    saveDestinationForAllStub = jest.fn();
    showDialogStub = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('render', () => {
    it('should render SubHeader with title Passport', () => {
      const { getByText } = createPageComponent();

      expect(getByText('Travel Document')).toBeInTheDocument();
    });

    it('should pass documentTitles to AdditionalPassportInfoForm when provided', () => {
      const documentTitles = { destination: 'Travel Destination' };
      const { container } = createPageComponent({ documentTitles });

      expect(container).toMatchSnapshot();
    });

    describe('submit button text', () => {
      it('should be Confirm and Continue when isLastPage is false', () => {
        const { getByText } = createPageComponent({});

        expect(getByText('CHECK_IN__CONFIRM_AND_CONTINUE_BUTTON')).toBeInTheDocument();
      });

      it('should be Continue when isLastPage is true', () => {
        const { getByText } = createPageComponent({ isLastPAX: true });

        expect(getByText('SHARED__BUTTON_TEXT__CONTINUE')).toBeInTheDocument();
      });
    });

    describe('missing all additional info', () => {
      it('should init additional passport page with green card, visa and destination', () => {
        const { getByText } = createPageComponent({
          initialFormData: {
            destination: null,
            permanentResidentCard: null,
            visa: null
          }
        });

        expect(getByText('Destination Address')).toBeInTheDocument();
        expect(getByText('Green Card')).toBeInTheDocument();
        expect(getByText('Visa')).toBeInTheDocument();
      });
    });

    describe('Cancel button', () => {
      it('should render the Cancel button at left side in the form', () => {
        const { getAllByText } = createPageComponent({
          recordLocator: 'PNR123'
        });
        const cancelButton = getAllByText('Cancel')[0];

        expect(cancelButton).toBeDefined();
      });

      describe('Confirmation dialog been showed', () => {
        it('should render dialog correctly', () => {
          const { getAllByText } = createPageComponent({
            recordLocator: 'PNR123'
          });
          const cancelButton = getAllByText('Cancel')[0];

          fireEvent.click(cancelButton);

          expect(showDialogStub).toHaveBeenCalled();
          expect(showDialogStub.mock.calls[0][0]).toMatchSnapshot();
        });

        it('should transition to previous page when click Yes button', async () => {
          const { getAllByText } = createPageComponent({
            recordLocator: 'PNR123'
          });
          const cancelButton = getAllByText('Cancel')[0];

          fireEvent.click(cancelButton);

          await waitFor(() => {
            expect(showDialogStub.mock.calls[0][0].buttons).toHaveLength(2);
            showDialogStub.mock.calls[0][0].buttons[1].onClick();

            expect(hideDialogStub).toHaveBeenCalled();
            expect(goBackStub).toHaveBeenCalled();
          });
        });

        it('should not call transition to checkin detail when click No button', () => {
          const { getAllByText } = createPageComponent({
            recordLocator: 'PNR123'
          });
          const cancelButton = getAllByText('Cancel')[0];

          fireEvent.click(cancelButton);

          expect(showDialogStub.mock.calls[0][0].buttons).toHaveLength(2);
          showDialogStub.mock.calls[0][0].buttons[0].onClick();

          expect(hideDialogStub).toHaveBeenCalled();
        });
      });
    });

    describe('Skip button', () => {
      it('should render the Skip button at right side in the form when there is more than one passenger', () => {
        const { getByText } = createPageComponent({
          shouldShowSkipButton: true
        });

        expect(getByText('Skip')).toBeInTheDocument();
      });

      it('should not render the Skip button at right side in the form when there is only one passenger', () => {
        const { queryByText } = createPageComponent();

        expect(queryByText('Skip')).not.toBeInTheDocument();
      });

      it('should go to the passport page for the 2nd passenger when current pax is not checkInConfirmation', () => {
        const { container } = createPageComponent({
          nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInPassportInformation' }), nextPaxNumber: '2' },
          shouldShowSkipButton: true
        });
        const skipButton = container.querySelector('.action-bar--right-buttons button');

        fireEvent.click(skipButton);

        expect(pushStub).toHaveBeenCalledWith('/check-in/2/passportPage');
      });

      it('should transition to confirmation when current pax is the last one who need to update apis', () => {
        const { container } = createPageComponent({
          shouldShowSkipButton: true,
          travelDocuments: [
            {
              missingDocuments: ['VISA']
            }
          ]
        });
        const skipButton = container.querySelector('.action-bar--right-buttons button');

        fireEvent.click(skipButton);

        expect(pushStub).toHaveBeenCalledWith('/check-in/confirmation');
      });
    });

    describe('passenger name', () => {
      it('should re-render passenger name', () => {
        const givenPassengerName = 'Hello world';
        const { getByText } = createPageComponent({ passengerName: givenPassengerName });

        expect(getByText(givenPassengerName)).toBeInTheDocument();
      });
    });
  });

  describe('submit', () => {
    describe('form data is not empty', () => {
      let checkInSessionToken;
      let initialFormData;
      let requestData;

      beforeEach(() => {
        checkInSessionToken = 'checkInSessionToken';
        initialFormData = { visa: true };
        requestData = { href: 'url', body: {} };
      });

      it('should call addAdditionalPassportInfoDocuments action with correct params', () => {
        const { container } = createPageComponent({
          checkInSessionToken,
          initialFormData,
          nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformation' }), nextPaxNumber: '2' },
          requestData
        });
        const form = container.querySelector('form');

        fireEvent.submit(form);

        expect(addAdditionalPassportInfoDocumentsStub).toHaveBeenCalledWith(
          requestData,
          initialFormData,
          checkInSessionToken
        );
      });

      it('should call transition to when current pax is not checkInConfirmation', async () => {
        const { container } = createPageComponent({
          checkInSessionToken,
          initialFormData,
          nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInPassportInformation' }), nextPaxNumber: '2' },
          requestData
        });
        const form = container.querySelector('form');

        fireEvent.submit(form);

        await waitFor(() => {
          expect(pushStub).toHaveBeenCalledWith('/check-in/2/passportPage');
        });
      });

      describe('check in session token expired when save travel documents', () => {
        const error = {
          $customized: true,
          responseJSON: { code: 400511157 }
        };

        beforeEach(() => {
          addAdditionalPassportInfoDocumentsStub.mockReturnValueOnce(Promise.reject(error));
        });

        it('should show error pop with session token expired message', async () => {
          const { container } = createPageComponent({
            checkInSessionToken,
            initialFormData,
            nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformation' }), nextPaxNumber: '2' },
            requestData
          });
          const form = container.querySelector('form');

          fireEvent.submit(form);

          await waitFor(() => {
            expect(showDialogStub).toHaveBeenCalled();
            expect(showDialogStub.mock.calls[0][0].name).toBe('check-in-travel-documents-session-token-expired');
            expect(showDialogStub.mock.calls[0][0].title).toBe('CHECK_IN__ERRORS__SESSION_TOKEN_EXPIRED');
            expect(showDialogStub.mock.calls[0][0].buttons[0].label).toBe('SHARED__BUTTON_TEXT__OK');
          });
        });

        it('should refresh app when user click ok button on error popup', async () => {
          const { container } = createPageComponent({
            checkInSessionToken,
            initialFormData,
            nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInAdditionalPassportInformation' }), nextPaxNumber: '2' },
            requestData
          });

          fireEvent.submit(container.querySelector('form'));

          await waitFor(() => {
            showDialogStub.mock.calls[0][0].buttons[0].onClick();

            expect(showDialogStub.mock.calls[0][0].buttons).toHaveLength(1);
            expect(hideDialogStub).toHaveBeenCalled();
            expect(BrowserObject.location.reload).toHaveBeenCalled();
          });
        });

        it('should logout user when user click ok button on error popup and login session also expired', async () => {
          jest.spyOn(LoginSessionHelper, 'hasSessionExpired').mockReturnValue(true);
          const state = { app: { account: { isLoggedIn: true } } };
          
          const { container } = createPageComponent({}, state);
          const form = container.querySelector('form');

          fireEvent.submit(form);

          await waitFor(() => {
            expect(showDialogStub.mock.calls[0][0].buttons).toHaveLength(1);
            showDialogStub.mock.calls[0][0].buttons[0].onClick();

            expect(cleanUpEndOfSessionFnStub).toHaveBeenCalled();
            expect(hideDialogStub).toHaveBeenCalled();
          });
        });
      });
    });
  });

  describe('onClick', () => {
    it('should transition to permanentResidentCard page when click the permanent resident card nav item', () => {
      const { container } = createPageComponent({
        initialFormData: { permanentResidentCard: null }
      });
      const permanentResidentCardNavItem = container.querySelector(`a[name='permanentResidentCard']`);

      fireEvent.click(permanentResidentCardNavItem);

      expect(pushStub).toHaveBeenCalledWith('/check-in/1/additional-passport-info/green-card');
    });
  });

  const createPageComponent = (props = {}) => {
    const defaultProps = {
      addAdditionalPassportInfoDocumentsFn: addAdditionalPassportInfoDocumentsStub,
      checkInSessionToken: '',
      cleanUpEndOfSessionFn: cleanUpEndOfSessionFnStub,
      goBack: goBackStub,
      hideDialogFn: hideDialogStub,
      initialFormData: {},
      isLastPAX: false,
      isLoggedIn: true,
      location: { search: 'search' },
      nextPageOptions: { nextPagePath: getNormalizedRoute({ routeName: 'checkInConfirmation' }), nextPaxNumber: '1' },
      params: { paxNumber: 1 },
      passengerName: '',
      prefillNextPaxInfoFn: prefillNextPaxInfoStub,
      prepareTravelDocumentFn: prepareTravelDocumentStub,
      push: pushStub,
      recordLocator: '',
      requestData: {},
      saveDestinationForAllFn: saveDestinationForAllStub,
      shouldShowSkipButton: false,
      showDialogFn: showDialogStub
    };
    const updatedProps = { ...defaultProps, ...props };

    return integrationRender({ withDialog: true })({}, AdditionalPassportInfoPage, updatedProps);
  };
});
