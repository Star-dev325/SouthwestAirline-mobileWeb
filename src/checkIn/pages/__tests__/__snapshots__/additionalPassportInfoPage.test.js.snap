// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdditionalPassportInfoPage render Cancel button Confirmation dialog been showed should render dialog correctly 1`] = `
Object {
  "buttons": Array [
    Object {
      "label": "SHARED__BUTTON_TEXT__NO",
      "onClick": [MockFunction],
    },
    Object {
      "label": "SHARED__BUTTON_TEXT__YES",
      "onClick": [Function],
    },
  ],
  "message": "SHARED__PASSPORT_PAGE__CANCEL_DIALOG_MESSAGE_FOR_CHECK_IN",
  "name": "addition-quit-apis-flow",
  "title": "SHARED__PASSPORT_PAGE__CANCEL_DIALOG_TITLE",
}
`;

exports[`AdditionalPassportInfoPage render should pass documentTitles to AdditionalPassportInfoForm when provided 1`] = `
<div>
  <div>
    <div>
      <div />
      <div>
        <div>
          <div
            class="page-header action-bar"
            id=""
          >
            <ul
              class="action-bar-buttons action-bar--left-buttons"
            >
              <li
                class="action-bar-buttons--item"
              >
                <button
                  class="button"
                  type="button"
                >
                  Cancel
                </button>
              </li>
            </ul>
            <div
              class="action-bar--container"
            >
              <div
                class="action-bar--title-align-center action-bar--title"
              >
                <span
                  class="action-bar--main-title"
                >
                  Travel Document
                </span>
                <span
                  class="action-bar--sub-title"
                />
              </div>
            </div>
            <ul
              class="action-bar-buttons action-bar--right-buttons"
            />
          </div>
          <div />
        </div>
        <form
          class="passport-form form"
          name="CHECK_IN_ADDITIONAL_PASSPORT_INFO_FORM_1"
          novalidate=""
        >
          <fieldset>
            <div
              class="passport-form--passenger"
            >
              <p
                class="passport-form--passenger-label"
              >
                SHARED__PLACEHOLDER__PASSENGER
              </p>
              <span
                class="passport-form--passenger-name"
              />
            </div>
            <div
              class="segments"
            >
              <div
                class="passport-form--info secondary segment"
              >
                <div
                  class="gray5"
                >
                  SHARED__PLACEHOLDER__TRAVEL_DOCUMENT_INFORMATION
                </div>
              </div>
              <div
                class="segment"
              >
                <button
                  class="check-in button button--fluid huge button--yellow"
                  type="submit"
                >
                  CHECK_IN__CONFIRM_AND_CONTINUE_BUTTON
                </button>
              </div>
            </div>
          </fieldset>
        </form>
      </div>
    </div>
  </div>
</div>
`;
